### PR TITLE
Translate CAST to connector expression

### PR DESCRIPTION
**Pull request by wendigo**
Extracted first commit from https://github.com/trinodb/trino/pull/11522

### CODE REVIEW

**Comment by wendigo** [comment]
@findepi PTAL

**Comment by wendigo** [comment]
Rebased to include https://github.com/trinodb/trino/pull/11567

**Comment by findepi** [comment]
CI failure looks related.

**Comment by wendigo** [comment]
@findepi fixed

**Comment by findepi** [comment]
CI: https://github.com/trinodb/trino/issues/11275

**Comment by findepi** [comment]
CAST has the property that the target type is known, so it probably shouldn't be `ConnectorExpression`.
If it is, there is no way to validate it's correctness.

I think we should model the CAST as a `Cast` class, new type of `ConnectorExpression`.
I will talk this with Martin

cc @martint @hashhar @losipiuk @assaf2

**Comment by wendigo** [comment]
@findepi ptal

**Comment by findepi** [comment]
I discussed this with @martint  yesterday, and the conclusion was to model the CAST as a 1-arg `$cast` function.

The concerns were that it would prevent type inference if it is needed in the future, but we believe it won't be needed nor desired anyway. Also, our ConnectorExpression "language" doesn't have implicit coercions, and we don't want to change this.

File: core/trino-spi/src/main/java/io/trino/spi/expression/StandardFunctions.java [hunk 1]
```diff
@@ -38,6 +38,8 @@ private StandardFunctions() {}
      */
     public static final FunctionName NULLIF_FUNCTION_NAME = new FunctionName("$nullif");
 
+    public static final FunctionName CAST_FUNCTION_NAME = new FunctionName("$cast");
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/expression/StandardFunctions.java:41]: add javadoc, indicating that the result type is given by `{@link Call}`'s type

File: core/trino-main/src/test/java/io/trino/sql/planner/TestTableScanRedirectionWithPushdown.java [hunk 2]
```diff
@@ -77,7 +77,8 @@
     private static final String TEST_TABLE = "test_table";
     private static final SchemaTableName SOURCE_TABLE = new SchemaTableName(TEST_SCHEMA, TEST_TABLE);
 
-    private static final Session MOCK_SESSION = testSessionBuilder().setCatalog(MOCK_CATALOG).setSchema(TEST_SCHEMA).build();
+    // Disable complex expression pushdown so CAST are not translated
+    private static final Session MOCK_SESSION = testSessionBuilder().setCatalog(MOCK_CATALOG).setSchema(TEST_SCHEMA).setSystemProperty("complex_expression_pushdown", "false").build();
```
Comment (findepi) [inline core/trino-main/src/test/java/io/trino/sql/planner/TestTableScanRedirectionWithPushdown.java:81]: i don't think we want this.

- complex_expression_pushdown disables more than just CASTs
- complex_expression_pushdown=true is the default, and we want to first and foremost rigorously test the default setup

File: core/trino-main/src/test/java/io/trino/sql/planner/TestConnectorExpressionTranslator.java [hunk 3]
```diff
@@ -298,6 +301,18 @@ public void testTranslateNullIf()
...
     }
 
+    @Test
+    public void testTranslateCast()
+    {
+        assertTranslationRoundTrips(
+                new Cast(
+                        new SymbolReference("varchar_symbol_1"),
+                        toSqlType(VARCHAR_TYPE)),
+                new Call(VARCHAR_TYPE,
```
Comment (findepi) [inline core/trino-main/src/test/java/io/trino/sql/planner/TestConnectorExpressionTranslator.java:311]:
```suggestion
                new Call(
                    VARCHAR_TYPE,
```

File: core/trino-main/src/test/java/io/trino/sql/planner/TestConnectorExpressionTranslator.java [hunk 4]
```diff
@@ -298,6 +301,18 @@ public void testTranslateNullIf()
                                 new Variable("varchar_symbol_1", VARCHAR_TYPE))));
     }
 
+    @Test
+    public void testTranslateCast()
```
Comment (findepi) [inline core/trino-main/src/test/java/io/trino/sql/planner/TestConnectorExpressionTranslator.java:305]: add a test case with a type-only cast

File: core/trino-main/src/test/java/io/trino/sql/planner/TestTableScanRedirectionWithPushdown.java (hunk 2 above)
Comment (wendigo) [reply core/trino-main/src/test/java/io/trino/sql/planner/TestTableScanRedirectionWithPushdown.java:81]: complex_expression_pushdown breaks this test and I don't know how to correctly fix it

Comment (findepi) [reply core/trino-main/src/test/java/io/trino/sql/planner/TestTableScanRedirectionWithPushdown.java:81]: let's revert here and iterate from there

File: core/trino-main/src/test/java/io/trino/sql/planner/TestConnectorExpressionTranslator.java (hunk 4 above)
Comment (wendigo) [reply core/trino-main/src/test/java/io/trino/sql/planner/TestConnectorExpressionTranslator.java:305]: Done.

File: core/trino-main/src/test/java/io/trino/sql/planner/TestTableScanRedirectionWithPushdown.java (hunk 2 above)
Comment (wendigo) [reply core/trino-main/src/test/java/io/trino/sql/planner/TestTableScanRedirectionWithPushdown.java]: @raunaqmorarka can you help with that test?

Comment (raunaqmorarka) [reply core/trino-main/src/test/java/io/trino/sql/planner/TestTableScanRedirectionWithPushdown.java]: I think you need to modify the MockConnector setup in `createLocalQueryRunner` to not pushdown CAST. I'm guessing `mockApplyProjection` needs to be changed for that.

Comment (raunaqmorarka) [reply core/trino-main/src/test/java/io/trino/sql/planner/TestTableScanRedirectionWithPushdown.java]: I got it to work at https://github.com/raunaqmorarka/presto/commit/2316e60e746f24f12f50c52a1ac648462f8a55ae

File: core/trino-main/src/main/java/io/trino/sql/planner/ConnectorExpressionTranslator.java [hunk 5]
```diff
@@ -573,6 +590,21 @@ public SqlToConnectorExpressionTranslator(Session session, Map<NodeRef<Expressio
             if (isEffectivelyLiteral(plannerContext, session, node)) {
                 return Optional.of(constantFor(node));
             }
+
+            if (node.isSafe()) {
+                return Optional.empty();
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/planner/ConnectorExpressionTranslator.java:595]: Add explanatory comment

**Review by findepi — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/sql/planner/ConnectorExpressionTranslator.java (hunk 5 above)
Comment (findepi) [reply core/trino-main/src/main/java/io/trino/sql/planner/ConnectorExpressionTranslator.java:596]: added

**Comment by wendigo** [comment]
```
2022-04-07T13:29:57.5906909Z org.testng.internal.thread.ThreadTimeoutException: Method io.trino.execution.resourcegroups.db.TestQueuesDb.testSelectorPriority() didn't finish within the time-out 60000
```

**Pull request merged**